### PR TITLE
Add create post static page, Add createPost REST API(tested with Postman)

### PR DESCRIPTION
**Pull request by isabella0428**
Create static `create post` widget based on wechat's design.
createPost

### CODE REVIEW

**Comment by isabella0428** [comment]
Just look at the commit `67cf0ac`, I revert the previous commit since it involves `pubspec.lock`

File: flutter/Graminsta/lib/createPostWidget.dart [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+class CreatePostWidget extends StatefulWidget {
+  CreatePostWidget({Key key}) : super(key: key);
+
+  @override
+  _CreatePostWidgetState createState() => _CreatePostWidgetState();
+}
+
+class _CreatePostWidgetState extends State<CreatePostWidget> {
+  String accessSetting = "All";
```
Comment (Time1ess) [inline flutter/Graminsta/lib/createPostWidget.dart]: I'd suggest to use a enum instead.

```
enum AccessControlType {
  public,
  private,
  visibleToSpecificFriends,
  invisibleToSpecificFriends,
}
```

and maybe a helper function to return `String` based on the enum.

```
extention AccessControlTypeExtension on AccessControlType {
  String toHumanReadableString() {
    switch(this):
      case AccessControlType.public:
        return "ALL";
  ...
}
```

See also: https://stackoverflow.com/questions/38908285/add-methods-or-values-to-enum-in-dart

File: flutter/Graminsta/lib/createPostWidget.dart [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+          IconButton(
+            icon: Icon(
+              Icons.save,
+              color: Colors.white,
+            ),
+            onPressed: () {
+              // do something
+            },
+          )
+        ]),
```
Comment (Time1ess) [inline flutter/Graminsta/lib/createPostWidget.dart]: Add trailing commas for all function calls and definitions, ditto for all.

Exception: function call or definition, which can be fitted in one line.

Comment (Time1ess) [reply flutter/Graminsta/lib/createPostWidget.dart]: One line: 80 chars.

Comment (Time1ess) [reply flutter/Graminsta/lib/createPostWidget.dart]: @shiyi001 @hht-hannah @ArianaGu FYI

**Review by Time1ess — CHANGES_REQUESTED** [review]
Can we reduce the size of the camera button?

**Comment by isabella0428** [comment]
> Can we reduce the size of the camera button?

It is not intended just to be a button.After we click the camera button, we will upload a photo and its thumbnail will be displayed in the button area, like wechat.

File: flutter/Graminsta/lib/createPostWidget.dart (hunk 2 above)
Comment (isabella0428) [reply flutter/Graminsta/lib/createPostWidget.dart]: > One line: 80 chars.

I used flutter format to form the file, does it exceeds the limit?

**Comment by isabella0428** [comment]
Add createPost rest API @ @Time1ess @shiyi001 @hht-hannah @ArianaGu

File: flutter/Graminsta/lib/createPostWidget.dart (hunk 2 above)
Comment (Time1ess) [reply flutter/Graminsta/lib/createPostWidget.dart]: As per https://dart.dev/guides/language/effective-dart/style#avoid-lines-longer-than-80-characters, you should have at most 80 chars per line.

@shiyi001 @hht-hannah @ArianaGu FYI

File: backend/graminsta/core/migrations/0001_initial.py [hunk 3]
```diff
@@ -1,26 +0,0 @@
-# Generated by Django 3.0.8 on 2020-07-30 06:07
```
Comment (Time1ess) [inline backend/graminsta/core/migrations/0001_initial.py]: Why is this deleted? Try to restore it.

File: backend/graminsta/core/urls.py [hunk 4]
```diff
@@ -5,7 +5,7 @@
 """
 
 from django.urls import path
-from core.views import UserInfoRecordView
+from .views import UserInfoRecordView
```
Comment (Time1ess) [inline backend/graminsta/core/urls.py]: Don't change this.

File: backend/graminsta/graminsta/settings.py [hunk 5]
```diff
@@ -25,6 +25,9 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = True
 
+# Root path for resources uploaded
+MEDIA_ROOT = "img/"
```
Comment (Time1ess) [inline backend/graminsta/graminsta/settings.py]: As there are multiple types of media (image, text, video, audio), use `media/` as `MEDIA_ROOT` might be better?

File: backend/graminsta/post/models.py [hunk 6]
```diff
@@ -12,9 +13,12 @@ class Post(models.Model):
     publisher = models.ForeignKey(auth_models.User, on_delete=models.CASCADE,
                                   related_name="published_posts")
     description = models.TextField()
-    img = models.ImageField(upload_to='img')
+    location = models.TextField()
```
Comment (Time1ess) [inline backend/graminsta/post/models.py]: I think maybe you can use two `double` fields for lng, lat instead?

File: backend/graminsta/post/services.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+Service functions for post module
+"""
+
+import datetime
+from django.contrib.auth import get_user_model
+from .models import Post
+
+
+def create_post(publisher_id, description, img):
+    """Create or update a Post in datebase.
```
Comment (Time1ess) [inline backend/graminsta/post/services.py]: The documentation doesn't match the method name? `create or update` vs `create`. And do separate `create_post` and `update_post`.

File: backend/graminsta/post/services.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    publisher_name: string
+        name of the publisher
+    description: string
+        description of the post
+    img:  binary data
+        image content of the post
+
+    Returns
+    -------
+    post: json format
```
Comment (Time1ess) [inline backend/graminsta/post/services.py]: Return `Post` object instead of JSON.

File: backend/graminsta/post/services.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
+"""
+Service functions for post module
+"""
+
+import datetime
+from django.contrib.auth import get_user_model
+from .models import Post
+
+
+def create_post(publisher_id, description, img):
```
Comment (Time1ess) [inline backend/graminsta/post/services.py]: I think to create a `Post`, you need `publisher_id`, `description`, `mention_user`, `image`

File: backend/graminsta/post/urls.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+Urls for post app.
+"""
+
+from django.urls import path
+from .views import PostRecordView
+
+
+app_name = 'post'
+urlpatterns = [
+    path('createPost/', PostRecordView.as_view(), name='post'),
```
Comment (Time1ess) [inline backend/graminsta/post/urls.py]: Don't use `createPost`, instead, use `create-post`. As per Google search guidelines:

> Consider using punctuation in your URLs. The URL http://www.example.com/green-dress.html is much more useful to us than http://www.example.com/greendress.html. We recommend that you use hyphens (-) instead of underscores (_) in your URLs.

See also: https://support.google.com/webmasters/answer/76329?hl=en

File: backend/graminsta/post/views.py [hunk 11]
```diff
@@ -1,3 +1,42 @@
...
+            Data containing publisher_id, description and image binary data
+
+        Returns
+        ----------
+        response: json format
+            Newly created post
+        """
+
+        # _ is not allowed in header key
+        publisher_id = int(request.META.get("HTTP_PUBLISHERID"))
```
Comment (Time1ess) [inline backend/graminsta/post/views.py:92]: I think to get current user, you can use `request.user` (if authentication works, otherwise you can leave a TODO here)`

File: flutter/Graminsta/lib/createPostWidget.dart [hunk 12]
```diff
@@ -0,0 +1,129 @@
+import 'package:flutter/material.dart';
```
Comment (Time1ess) [inline flutter/Graminsta/lib/createPostWidget.dart:1]: Usually, to reduce the burden for reviewers, try to split code into at least backend and frontend. But do that in following PRs, not this one.

**Review by Time1ess — CHANGES_REQUESTED** [review]

File: backend/graminsta/post/views.py (hunk 11 above)
Comment (isabella0428) [reply backend/graminsta/post/views.py:92]: Add TODO here

File: backend/graminsta/post/services.py (hunk 7 above)
Comment (isabella0428) [reply backend/graminsta/post/services.py]: Changed to create

File: backend/graminsta/post/models.py (hunk 6 above)
Comment (isabella0428) [reply backend/graminsta/post/models.py]: Changed to lat and lng

File: backend/graminsta/post/services.py (hunk 9 above)
Comment (isabella0428) [reply backend/graminsta/post/services.py]: Added

File: flutter/Graminsta/lib/createPostWidget.dart (hunk 2 above)
Comment (isabella0428) [reply flutter/Graminsta/lib/createPostWidget.dart]: I set a ruler here and it seems not exceeding the 80 chars limit.
ruler

File: __azurite_db_queue__.json [hunk 13]
```diff
@@ -0,0 +1 @@
...},"queueName":{"name":"queueName","dirty":false,"values":[]},"messageId":{"name":"messageId","dirty":false,"values":[]},"visibleTime":{"name":"visibleTime","dirty":false,"values":[]}},"constraints":null,"uniqueNames":[],"transforms":{},"objType":"$MESSAGES_COLLECTION$","dirty":false,"cachedIndex":null,"cachedBinaryIndex":null,"cachedData":null,"adaptiveBinaryIndices":true,"transactional":false,"cloneObjects":false,"cloneMethod":"parse-stringify","asyncListeners":false,"disableMeta":false,"disableChangesApi":true,"disableDeltaChangesApi":true,"autoupdate":false,"serializableIndices":true,"ttl":null,"maxId":0,"DynamicViews":[],"events":{"insert":[],"update":[],"pre-insert":[],"pre-update":[],"close":[],"flushbuffer":[],"error":[],"delete":[null],"warning":[null]},"changes":[]}],"databaseVersion":1.5,"engineVersion":1.5,"autosave":true,"autosaveInterval":5000,"autosaveHandle":null,"throttledSaves":true,"options":{"autosave":true,"autosaveInterval":5000,"serializationMethod":"normal","destructureDelimiter":"$<\n"},"persistenceMethod":"fs","persistenceAdapter":null,"verbose":false,"events":{"init":[null],"loaded":[],"flushChanges":[],"close":[],"changes":[],"warning":[]},"ENV":"NODEJS"}
```
Comment (Time1ess) [inline __azurite_db_queue__.json]: Ignore this file?

File: backend/graminsta/post/models.py [hunk 14]
```diff
@@ -12,9 +13,13 @@ class Post(models.Model):
     publisher = models.ForeignKey(auth_models.User, on_delete=models.CASCADE,
                                   related_name="published_posts")
     description = models.TextField()
-    img = models.ImageField(upload_to='img')
+    lng = models.FloatField(default=100)
```
Comment (Time1ess) [inline backend/graminsta/post/models.py]: I think you don't need default value. Maybe `blank=True` and `null=True` can help?

```
lon = models.FloatField(blank=True, null=True)
lat = models.FloatField(blank=True, null=True)
```

File: backend/graminsta/post/services.py [hunk 15]
```diff
@@ -21,17 +21,21 @@ def create_post(publisher_id, description, img):
 
     Returns
     -------
-    post: json format
-        a successfully created post
+    post: a successfully created post object
     """
     # Get publisher object with publisher_name
     publisher = get_user_model().objects.get(pk=publisher_id)
+    print(publisher.id)
```
Comment (Time1ess) [inline backend/graminsta/post/services.py]: Remove print.

File: backend/graminsta/post/services.py [hunk 16]
```diff
@@ -21,17 +21,21 @@ def create_post(publisher_id, description, img):
...
         publisher=publisher,
         description=description,
         img=img,
-        created_at=datetime.datetime.now()
+        created_at=datetime.datetime.now(),
     )
+
+    user_ids = mention_user_ids.split(",")
+    for user_id in user_ids:
+        post.mention_user.add(get_user_model().objects.get(pk=user_id))
```
Comment (Time1ess) [inline backend/graminsta/post/services.py]: I think you can get all users in one shot.

```
users = get_user_model().objects.filter(pk__in=user_ids)
```

File: backend/graminsta/post/urls.py [hunk 17]
```diff
@@ -8,5 +8,5 @@
 
 app_name = 'post'
 urlpatterns = [
-    path('createPost/', PostRecordView.as_view(), name='post'),
+    path('create-post/', PostRecordView.as_view(), name='post'),
```
Comment (Time1ess) [inline backend/graminsta/post/urls.py]: I think in your case, RESTful API `/<prefix>/post` can fit the requirements.

```
GET /<prefix>/post   Retrieve list of Posts
GET /<prefix>/post/5  Retrieve Post whose id is 5
POSt /<prefix>/post    Create a Post
```

Then your url endpoints need to be updated.

**Review by Time1ess — CHANGES_REQUESTED** [review]

File: backend/graminsta/graminsta/urls.py [hunk 18]
```diff
@@ -20,6 +20,7 @@
 urlpatterns = [
     path('admin/', admin.site.urls),
     path('core/', include('core.urls', namespace='core')),
+    path('post/', include('post.urls', namespace='post')),
     path('authtoken/', views.obtain_auth_token, name='authtoken'),
     path('post/', include('post.urls', namespace='post')),
```
Comment (Time1ess) [inline backend/graminsta/graminsta/urls.py]: We already have one `post/` from others' PR, please remove the duplicated one

File: backend/graminsta/post/services.py [hunk 19]
```diff
@@ -1,9 +1,44 @@
...
+    post: a successfully created post object
+    """
+    # Get publisher object with publisher_name
+    publisher = get_user_model().objects.get(pk=publisher_id)
+    # Get the default image
+    post = Post.objects.create(
+        publisher=publisher,
+        description=description,
+        img=img,
+        created_at=datetime.datetime.now(),
```
Comment (Time1ess) [inline backend/graminsta/post/services.py]: Since `created_at` has `auto_now_add=True`, Django will auto populate this field for you when you create the `Post`, so you don't need to pass this.

**Review by Time1ess — APPROVED** [review]

**Review by shiyi001 — APPROVED** [review]
